### PR TITLE
Added completion for unstable features

### PR DESCRIPTION
**Pull request by Fihtangolz**
Added xtask for downloading list of unstable features from the unstable book and codegen for it. Also included small changes from linter.

### CODE REVIEW

**Comment by Fihtangolz** [comment]
> Neat idea!
> 
> I wonder what the best way to keep this up-to-date would be though, since feature flags tend to change relatively often. I think CI checks that all codegen'd files are up-to-date, so that would start failing whenever nightly features are added/removed.
The best option would be for the CI to do the check itself and do codegene, but I'm not sure how this will affect the run time.

File: xtask/src/codegen/gen_unstable_future_descriptor.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    Command::new("git").current_dir(path.clone()).arg("init").output()?;
+    Command::new("git")
+        .current_dir(path.clone())
+        .args(&["remote", "add", "-f", "origin", codegen::REPOSITORY_URL])
+        .output()?;
+    Command::new("git")
+        .current_dir(path.clone())
+        .args(&["sparse-checkout", "set", "/src/doc/unstable-book/src/"])
+        .output()?;
```
Comment (bjorn3) [inline xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: https://www.git-scm.com/docs/git-sparse-checkout

> THIS COMMAND IS EXPERIMENTAL. ITS BEHAVIOR, AND THE BEHAVIOR OF OTHER COMMANDS IN THE PRESENCE OF SPARSE-CHECKOUTS, WILL LIKELY CHANGE IN THE FUTURE.

Comment (Fihtangolz) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: Well, yes, but how else can I get a separate directory and check it? I am really not good enough at git, so any ideas are welcome

Comment (bjorn3) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: A bit more complex, but you could use the github api. This also completely avoids the clone.

Comment (Fihtangolz) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: Ok, it's seems reasonable, I try it.

Comment (Fihtangolz) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: > A bit more complex, but you could use the github api. This also completely avoids the clone.
We need a token, what about it?

Comment (bjorn3) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: As far as I know you can access most apis without token, just with much lower rate limits.

https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api#authentication

> Unauthenticated clients can make 60 requests per hour. To get more requests per hour, we'll need to authenticate. In fact, doing anything interesting with the GitHub API requires authentication.

Comment (Fihtangolz) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: This is the problem, you can do a check, but  you cannot integrate this without a token in CI

Comment (bjorn3) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: CI can use `${{ secrets.GITHUB_TOKEN }}`: https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token

Comment (Fihtangolz) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs:22]: > CI can use `${{ secrets.GITHUB_TOKEN }}`: https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token

thank you for help

File: xtask/src/codegen/gen_unstable_future_descriptor.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    Command::new("git").current_dir(path.clone()).args(&["pull", "origin", "master"]).output()?;
```
Comment (matklad) [inline xtask/src/codegen/gen_unstable_future_descriptor.rs:23]: We have `not_bash` module for this kind of scripting:

https://github.com/rust-analyzer/rust-analyzer/blob/master/xtask/src/not_bash.rs

```
let _d = pushd(path);
run!("git pull origin master")?
...
```

File: xtask/src/codegen/gen_unstable_future_descriptor.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        })
+        .collect::<Vec<_>>();
+
+    let definitions = files
+        .iter()
+        .map(|entry| {
+            let path = entry.path();
+            let feature_ident =
+                format!("{}", path.file_stem().unwrap().to_str().unwrap().replace("-", "_"));
+            let doc = format!("{}", std::fs::read_to_string(path).unwrap());
```
Comment (matklad) [inline xtask/src/codegen/gen_unstable_future_descriptor.rs:43]: `not_bash` has `fs2` module, which gives better error messages.

File: crates/ra_ide/src/completion/complete_attribute.rs [hunk 4]
```diff
@@ -248,9 +258,9 @@ const DEFAULT_DERIVE_COMPLETIONS: &[DeriveCompletion] = &[
     DeriveCompletion { label: "Ord", dependencies: &["PartialOrd", "Eq", "PartialEq"] },
 ];
 
-struct LintCompletion {
-    label: &'static str,
-    description: &'static str,
+pub struct LintCompletion {
```
Comment (matklad) [inline crates/ra_ide/src/completion/complete_attribute.rs:261]: THis should be `pub(crate)` I think

File: xtask/src/codegen/gen_unstable_future_descriptor.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    let src_dir = path.join("src/doc/unstable-book/src");
+    let files = WalkDir::new(src_dir.join("language-features"))
+        .into_iter()
+        .chain(WalkDir::new(src_dir.join("library-features")))
+        .filter_map(|e| e.ok())
+        .filter(|entry| {
+            // Get all `.md ` files
+            entry.file_type().is_file()
+                && entry.path().extension().map(|ext| ext == "md").unwrap_or(false)
```
Comment (matklad) [inline xtask/src/codegen/gen_unstable_future_descriptor.rs:33]:
```suggestion
                && entry.path().extension().unwrap_or_default() == "md"
```

**Comment by matklad** [comment]
> I think CI checks that all codegen'd files are up-to-date, so that would start failing whenever nightly features are added/removed.

I'd rather not do this -- this'll add hard dependency on upstream,  which is bad for long-term project maintainance. 

CI can be broken only by changes to the files in this repository is the global property we want to enforce.

File: xtask/Cargo.toml [hunk 6]
```diff
@@ -18,3 +18,8 @@ quote = "1.0.2"
 ungrammar = "1.1.1"
 walkdir = "2.3.1"
 write-json =  "0.1.0"
+reqwest = { version = "0.10.7", features = ["blocking", "json"] }
+graphql_client = "0.9.0"
+serde = "1.0.115"
+regex = "1.3.9"
+chrono = "0.4.13"
```
Comment (matklad) [inline xtask/Cargo.toml:25]: xtask should have the least amount of deps possible.

Let's not do premature optimization and stick with boring `git clone https://github.com/rust-lang/rust/`

**Comment by Fihtangolz** [comment]
> > I think CI checks that all codegen'd files are up-to-date, so that would start failing whenever nightly features are added/removed.
> 
> I'd rather not do this -- this'll add hard dependency on upstream, which is bad for long-term project maintainance.
> 
> CI can be broken only by changes to the files in this repository is the global property we want to enforce.

Then there is no way to check the relevance and we just trust our contributors.  In general, I find being able to sync with parts of other repositories inevitable.

File: xtask/src/codegen/gen_unstable_future_descriptor.rs (hunk 1 above)
Comment (Fihtangolz) [reply xtask/src/codegen/gen_unstable_future_descriptor.rs]: mkladov said that this implementation is contrary to the goals

**Comment by matklad** [comment]
Needs a rebase

**Pull request merged**